### PR TITLE
Verify S/MIME signatures

### LINKED ISSUE #7666: Verify signed S/MIME messages
### Is your feature request related to a problem? Please describe.

As a user I want to see if an incoming message signed via S/MIME has a valid signature.

### Describe the solution you'd like

The app shall use the signature information attached to a message to verify if the contents have been tempered with. Moreover, the app shall check if the sender used a valid certificate. For infrastructure independence it could be wise to use Nextcloud's shipped CA bundle.

The trust model is to protect messages outside the boundaries of an organization. Therefore the server and its admins are trusted. Public and private keys may be known in plain text on the server-side.

The trust model is to protect messages outside the boundaries of an organization. Therefore the server and its admins are trusted. Public and private keys may be known in plain text on the server-side.

### Describe alternatives you've considered

N/a

### Additional context

This is part of https://github.com/nextcloud/mail/issues/7085.

### PR DESCRIPTION
**Pull request by st3iny**
Fix #7666 

| Verified signature | Unverified signature |
| --- | --- |
| image | image |

## TODO

- [x] Fix performance
  - [x] ~Verify signature in `\OCA\Mail\IMAP\MessageMapper::getBodyStructureData`? What about changing certificates (reverification, cache invalidation etc.)?~
  - [ ] Cache verification data like itineraries?
- [x] Implement indicator in frontend
- [x] Fix tests
- [x] Fix linter issues and static analysis
- [x] Get rid of openssl check as it is required anyway

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Do we need the info in the database? I guess that would only be a requirement if the envelope list receives some kind of icon, right? Otherweise we can do the verification at the time of fetching the message body, or async like itineraries as you suggested.

File: lib/Service/SMimeService.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	/**
+	 * Check whether the openssl module is available.
+	 * It provides the underlying S/MIME apis.
+	 *
+	 * @return bool
+	 */
+	public function hasOpenssl(): bool {
+		return extension_loaded('openssl');
```
Comment (kesselb) [inline lib/Service/SMimeService.php:50]: FYI: php-openssl is a hard requirement for Nextcloud (https://docs.nextcloud.com/server/latest/admin_manual/installation/source_installation.html). Nextcloud don't work without php-openssl :boom:.

Comment (st3iny) [reply lib/Service/SMimeService.php:50]: Thanks for letting me know. I'll get rid of the check.

File: lib/Controller/MessagesController.php [hunk 2]
```diff
@@ -219,6 +241,18 @@ public function getBody(int $id): JSONResponse {
...
 
+		$sMime = ['isSigned' => false];
+		try {
+			$parsedMessage = Horde_Mime_Part::parseMessage($rawMessage, ['no_body' => true]);
+			if ($parsedMessage->getType() === 'multipart/signed') {
+				$sMime['isSigned'] = true;
+				$sMime['signatureIsValid'] = $this->sMimeService->verifyMessage($rawMessage);
+			}
+		} catch (Horde_Mime_Exception $e) {
+			$this->logger->warning('Failed to parse MIME message', ['error' => $e]);
```
Comment (ChristophWurst) [inline lib/Controller/MessagesController.php:252]:
```suggestion
			$this->logger->warning('Failed to parse MIME message: ' . $e->getMessage(), ['error' => $e]);
```

Reveals the inner exception message when someone reports the bug

File: composer.json [hunk 3]
```diff
@@ -17,6 +17,7 @@
 	},
 	"require": {
 		"php": ">=7.4 <=8.1",
+		"ext-openssl": "*",
```
Comment (st3iny) [inline composer.json:20]: This was recommended by PHPStorm and can be reverted if necessary. According to the Nextcloud admin docs, openssl is required in order to run the server. This is more of a cosmetic thing.

File: lib/Controller/MessagesController.php [hunk 4]
```diff
@@ -219,6 +241,18 @@ public function getBody(int $id): JSONResponse {
 		$json['databaseId'] = $message->getId();
 		$json['isSenderTrusted'] = $this->isSenderTrusted($message);
 
+		$sMime = ['isSigned' => false];
```
Comment (ChristophWurst) [inline lib/Controller/MessagesController.php:244]: nit: make this a small serializable dto

**Review by ChristophWurst — APPROVED** [review]
Code looks good otherwise

On the UI I wouldn't mind if the lock icon was even colored yellow like we do with *important* messages, just to give a nice indication for a valid S/MIME. And red for invalid.

File: src/components/ThreadEnvelope.vue [hunk 5]
```diff
@@ -77,6 +77,16 @@
...
 				<template v-if="expanded">
+					<NcActions v-if="sMimeData.isSigned">
+						<template #icon>
+							<LockIcon v-if="sMimeData.signatureIsValid" :size="20" />
+							<LockOffIcon v-else :size="20" />
+						</template>
+						<NcActionText>
+							{{ sMimeSignMessage }}
+						</NcActionText>
+						<!-- TODO: display information about signer and/or CA certificate -->
```
[the remark this answers is not in the document]
Comment (st3iny) [reply src/components/ThreadEnvelope.vue:88]: Yeah, I wanted to get the basic functionality out asap.

[the remark this answers is not in the document]
Comment (st3iny) [reply src/components/ThreadEnvelope.vue:88]: Also our designers might also have an opinion on that because the extra information might be too technical.

**Comment by st3iny** [comment]
> On the UI I wouldn't mind if the lock icon was even colored yellow like we do with _important_ messages, just to give a nice indication for a valid S/MIME. And red for invalid.

Done. I updated both screenshots.

File: lib/Controller/MessagesController.php (hunk 4 above)
Comment (st3iny) [reply lib/Controller/MessagesController.php:244]: Done

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**